lib/Finance/Calendar.pm

```diff
@@ -585,30 +583,20 @@ Is this exchange trading on daylight savings times for the given epoch?
 
 ### PRIVATE ###
 
-has _holiday_cache => (
-    is      => 'rw',
-    default => sub { {} },
-);
-
 sub _get_holidays_for {
-    my ($self, $symbol) = @_;
+    my ($self, $symbol, $when) = @_;
 
-    my $cache = $self->_holiday_cache->{$symbol};
+    my $date     = $when->truncate_to_day->epoch;
+    my $calendar = $self->calendar->{holidays};
+    my $holiday  = $calendar->{$date};
 
-    return $cache if $cache;
+    return unless $holiday;
 
-    my $calendar = $self->calendar->{holidays};
-    my %holidays;
-    foreach my $date (keys %$calendar) {
-        foreach my $holiday_desc (keys %{$calendar->{$date}}) {
-            $holidays{Date::Utility->new($date)->days_since_epoch} = $holiday_desc
-                if (first { $symbol eq $_ } @{$calendar->{$date}{$holiday_desc}});
-        }
+    foreach my $holiday_desc (keys %$holiday) {
+        return $holiday_desc if (first { $symbol eq $_ } @{$holiday->{$holiday_desc}});
```

Comment:
Can there be multiple holiday descriptors for the same symbol? If so, you don't know which one you get here.

Also, it might be more efficient to store a reverse hash somewhere so that you don't have to search for the descriptor but can just `return $reverse_hash{$symbol}`. That, however, makes only sense if you don't build the hash each time before accessing it.